kaggle_environments/agent.py

```diff
@@ -12,23 +12,59 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import json
 import os
 import requests
-import json
-from multiprocessing import Manager, Process
+import sys
+from io import StringIO
 from time import time
-import uuid
-from .errors import DeadlineExceeded
-from .utils import get_exec, has, is_url, read_file, structify
+from urllib.parse import urlparse
+from .errors import DeadlineExceeded, InvalidArgument
+from .utils import read_file, structify
+
+
+def is_url(url):
+    try:
+        result = urlparse(url)
+        return all([result.scheme, result.netloc])
+    except ValueError:
+        return False
+
+
+def get_last_callable(raw, fallback=None):
+    orig_out = sys.stdout
+    buffer = StringIO()
+    sys.stdout = buffer
+
+    try:
+        code_object = compile(raw, "<string>", "exec")
+        env = {}
+        exec(code_object, env)
+        sys.stdout = orig_out
+        output = buffer.getvalue()
+        if output:
+            print(output)
+        return [v for v in env.values() if callable(v)][-1]
+    except Exception as e:
+        sys.stdout = orig_out
+        output = buffer.getvalue()
+        if output:
+            print(output)
+        if fallback is not None:
+            return fallback
+        raise InvalidArgument("Invalid raw Python: " + repr(e))
```

Comment:
Currently in most places the library uses `str(e)` to convert an exception to a string. I'm working on converting these to at least use `repr` which will include the name of the exception, but in all likelihood we want to eventually have a method that grabs and logs the stack frames as well.